### PR TITLE
Return HTTP 400 and a graphql error when registering an APQ operation with the wrong hash

### LINKED ISSUE #2948: APQ system should reject a query if its sha256Hash doesn't match
**Describe the bug**

The APQ docs for Apollo Server [state](https://www.apollographql.com/docs/apollo-server/performance/apq/#command-line-testing):

> The hash you provide _must_ be the exact SHA-256 hash of the query string. If it isn't, Apollo Server returns an error.

Bearing in mind that Apollo Server is not Apollo Router, I'd expect the same behavior from Router. However, Router only logs a warning in this situation, and allows the query to execute.

https://github.com/apollographql/router/blob/8d6f69d8266584f03474d817ac485faa2a5b8b97/apollo-router/src/services/layers/apq.rs#L83

**To Reproduce**
1. Issue a query to any Router that has APQ enabled, but specify an invalid sha256Hash. Example:

    ```sh
    curl --get https://www.example.com \
        --header "Content-Type: application/json" \
        --data-urlencode 'query={ viewer { username } }' \
        --data-urlencode 'extensions={"persistedQuery":{"version":1,"sha256Hash":"banana"}}'
    ```
2. Router will log a warning, but executes the query anyway and responds with the result.

**Expected behavior**

Router should reject the query with an error message/code, perhaps `PERSISTED_QUERY_NOT_FOUND`.

**Output**
n/a

**Desktop**
n/a

**Additional context**

In lieu of a good solution for query cost analysis, I wanted to use a Rhai script to create an allowlist of query hashes, similar to the request in #2528, but since Router permits queries that don't match the sha256Hash, this approach won't work.

<details>
  <summary>Expand to show the Rhai script</summary>

```rhai
let allowedQuerySha256Hashes = #{
    b0455b0dbfe6f76a7603e17cb9f0079fc6fc0286af0254a95462a06fa743757f: (),
    // ...
};

fn supergraph_service(service) {
    const request_callback = Fn("process_request");
    map_request(service, request_callback);
}

fn process_request(request) {
    if
        request.body.extensions == () ||
        request.body.extensions.persistedQuery == () ||
        request.body.extensions.persistedQuery.version != 1 ||
        !allowedQuerySha256Hashes.contains(request.body.extensions.persistedQuery.sha256Hash)
    {
        throw #{
            status: 403,
            body: #{
                errors: [#{
                    message: "Forbidden",
                    extensions: #{
                        code: "FORBIDDEN"
                    }
                }]
            }
        };
    }
}
```

</details>

### PR DESCRIPTION
**Pull request by o0Ignition0o**
Fixes #2948

Previously, when a client tried to persist a query with the wrong hash, we would log an error, and execute the query (without inserting the query into the APQ cache).
We now return a GraphQL error and don't execute the query.

This change reflects the gateway's behavior.

### CODE REVIEW

File: apollo-router/src/services/layers/apq.rs [hunk 1]
```diff
@@ -89,10 +90,32 @@ async fn apq_request(
...
+                Ok(request)
             } else {
-                tracing::warn!("apq: graphql request doesn't match provided sha256Hash");
+                tracing::debug!("apq: graphql request doesn't match provided sha256Hash");
+                let errors = vec![crate::error::Error {
+                    message: "provided sha does not match query".to_string(),
+                    locations: Default::default(),
+                    path: Default::default(),
+                    extensions: serde_json_bytes::from_value(json!({
+                          "code": "PERSISTED_QUERY_SHA_MISSMATCH",
```
Comment (abernix) [inline apollo-router/src/services/layers/apq.rs:101]: spelling.

```suggestion
                          "code": "PERSISTED_QUERY_SHA_MISMATCH",
```

File: apollo-router/src/services/layers/apq.rs [hunk 2]
```diff
@@ -89,10 +90,32 @@ async fn apq_request(
...
+                          "exception": {
+                          "stacktrace": [
```
Comment (abernix) [inline apollo-router/src/services/layers/apq.rs:103]: I'm not sure if this qualifies as a `stacktrace` in this case?

Comment (o0Ignition0o) [reply apollo-router/src/services/layers/apq.rs:103]: Indeed, I think I previously misread https://github.com/apollographql/apollo-server/blob/3d0d4b4b263e52bb36839221d9103f02ba1d6288/packages/integration-testsuite/src/httpServerTests.ts#L2672 and turned it into a `stacktrace`.

I'll update both errors (this one and persistedquerynotfound) to remove the stacktrace

File: apollo-router/src/services/layers/apq.rs (hunk 1 above)
Comment (o0Ignition0o) [reply apollo-router/src/services/layers/apq.rs]: fixed in aec6a223

File: apollo-router/src/services/layers/apq.rs (hunk 2 above)
Comment (o0Ignition0o) [reply apollo-router/src/services/layers/apq.rs]: fixed in aec6a223

**Review by abernix — COMMENTED** [review]
my initial feedback was addressed, leaving the full review/scrutiny up to the rest of the team.  thanks for picking this up!

**Review by BrynCooke — APPROVED** [review]

**Review by Geal — APPROVED** [review]

File: .changesets/fix_chirp_smoker_scout_carrier.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+### Return HTTP 400 and a graphql error when registering an APQ operation with the wrong hash ([Issue #2948](https://github.com/apollographql/router/issues/2948))
+
+Previously, when a client tried to persist a query with the wrong hash, we would log an error, and execute the query (without inserting the query into the APQ cache).
+We now return a GraphQL error and don't execute the query.
+
+This change reflects the gateway's behavior.
+
+By [@o0Ignition0o](https://github.com/o0Ignition0o) in https://github.com/apollographql/router/pull/3128
```
Comment (abernix) [inline .changesets/fix_chirp_smoker_scout_carrier.md:8]:
```suggestion
### An APQ query with a mismatched hash will error as HTTP 400 ([Issue #2948](https://github.com/apollographql/router/issues/2948))

We now model the both the behavior of the Gateway and the intended behavior of [the implementation](https://www.apollographql.com/docs/apollo-server/performance/apq/).  Even if our previous behavior was still acceptable, any other behavior is a misconfiguration of a client and should be prevented early.

Previously, if a client sent an operation with an APQ hash, we would merely log an error to the console, **not** register the operation (for the next request) but still execute the query.  We now return a GraphQL error and don't execute the query.  No clients should be impacted by this, though anyone who had hand-crafted a query **with** APQ information (for example, copied a previous APQ-registration query but only changed the operation without re-calculating the SHA-256) might now be forced to use the correct hash (or more practically, remove the hash).

By [@o0Ignition0o](https://github.com/o0Ignition0o) in https://github.com/apollographql/router/pull/3128
```

**Comment by Geal** [comment]
(disabling auto merge while we're doing the release dance) we should merge this though

**Comment by Geal** [comment]
we're out of the tricky release train, let's merge this

**Pull request merged**